.travis.yml

```diff
@@ -14,40 +14,87 @@ notifications:
-  - sh -e /etc/init.d/xvfb start
+  #- export CHROME_BIN=google-chrome-stable
+  #- export DISPLAY=:99.0
+  #- unset _JAVA_OPTIONS  # JVM heap sizes break closure compiler. #11203.
+  #- sh -e /etc/init.d/xvfb start
   - curl -o- -L https://yarnpkg.com/install.sh | bash
   - export PATH="$HOME/.yarn/bin:$HOME/.config/yarn/global/node_modules/.bin:$PATH"
-before_script:
-  - pip install --user protobuf
-script: node build-system/pr-check.js
-after_script:
-  - build-system/sauce_connect/stop_sauce_connect.sh
+  - pip install urllib3[secure]
+  - pip install gsutil --user
 branches:
   only:
     - master
     - release
     - canary
     - /^amp-release-.*$/
     - /^revert-.*$/
-addons:
-  chrome: stable
-  hosts:
-    - ads.localhost
-    - iframe.localhost
-    # Requested by some tests because they need a valid font host,
-    # but should not resolve in tests.
-    - fonts.googleapis.com
-  apt:
-    packages:
-      - protobuf-compiler
-      - python-protobuf
-matrix:
+stages:
+  - build
+  - test
+jobs:
+  fast_finish: true
+  allowed_failures:
+    - stage: test
+      name: "Remote (Sauce Labs) Tests"
+      script:
+        - node build-system/pr-check/remote-test.js
+      after_script:
+        - build-system/sauce_connect/stop_sauce_connect.sh
   include:
-    - env: BUILD_SHARD="unit_tests"
-    - env: BUILD_SHARD="integration_tests"
+    - stage: build
+      name: "Build"
+      script:
+        - node build-system/pr-check/build.js
+    - stage: build
+      name: "Checks" 
+      script:        
+        - node build-system/pr-check/checks.js
+    - stage: build
+      name: "Validator"        
+      before_script:
+        - pip install --user protobuf #required by validator
+      script:        
+        - node build-system/pr-check/validator.js
+      addons:
+        apt:
+          packages:
+            - protobuf-compiler
+            - python-protobuf
+    - stage: test
+      name: "Dist, Bundle Size, Single Pass Tests"
+      script:
+        - node build-system/pr-check/dist-test.js
+    - stage: test
+      name: "Visual Diff Tests"
+      script:
+        - node build-system/pr-check/visual-diff-test.js
+    - stage: test
+      name: "Local Tests"
+      before_install: 
+        - export CHROME_BIN=google-chrome-stable
+        - export DISPLAY=:99.0
+        # Explicitly add yarn and gsutil install because this section overrides before_install parent section
+        - curl -o- -L https://yarnpkg.com/install.sh | bash
+        - export PATH="$HOME/.yarn/bin:$HOME/.config/yarn/global/node_modules/.bin:$PATH"
+        - pip install urllib3[secure]
+        - pip install gsutil --user
+      addons:
+        chrome: stable
+        hosts:
+          - ads.localhost
+          - iframe.localhost
+          # Requested by some tests because they need a valid font host,
+          # but should not resolve in tests.
+          - fonts.googleapis.com
```

Comment:
If `hosts` is needed for one or more jobs, let's just move the entire `addons` section to the top to avoid repeating it everywhere. (It's inexpensive, and makes the rest of this file look cleaner.)